### PR TITLE
Two locus user-facing python ld matrix API

### PR DESCRIPTION
**Pull request by lkirk**
## Description

This is an implementation of the user-facing python API for `ld_matrix` as discussed in #2829

I've added some tests that cover most edge cases. I think I'd like to add some tests that store literal matricies, but I wanted to get eyes on this before I added those final tests.

# PR Checklist:

- [X] Tests that fully cover new/changed functionality.
- [ ] Documentation including tutorial content if appropriate.
- [ ] Changelogs, if there are API changes.

### CODE REVIEW

File: c/tskit/trees.c [hunk 1]
```diff
@@ -3638,9 +3637,10 @@ D_prime_summary_func(tsk_size_t state_dim, const double *state,
         double p_B = p_AB + p_aB;
 
         double D = p_AB - (p_A * p_B);
-        if (D >= 0) {
```
Comment (lkirk) [inline c/tskit/trees.c:3641]: The purpose of this change is to avoid dividing by zero. When D is zero, the denominator becomes zero.

File: python/tests/test_ld_matrix.py [hunk 2]
```diff
@@ -699,6 +735,210 @@ def r2_summary_func(
...
+) -> None:
+    """Summary function for the D statistic. We first compute the proportion of
+    AB, A, and B haplotypes, then we compute the D statistic, storing the outputs
+    in the result vector, one entry per sample set.
+
+    :param state_dim: Number of sample sets.
+    :param state: Counts of 3 haplotype configurations for each sample set.
+    :param result: Vector of length state_dim to store the results in.
+    :param params: Parameters for the summary function.
+    """
```
Comment (lkirk) [inline python/tests/test_ld_matrix.py:749]: Are these docstrings too much? I'm not sure how much they're really adding to the understanding of this code.

File: c/tskit/trees.c (hunk 1 above)
Comment (jeromekelleher) [reply c/tskit/trees.c:3641]: Should the result be NaN then? We should probably mark this as a follow-up issue, these things can be tricky to nail down.

File: python/tests/test_ld_matrix.py [hunk 3]
```diff
@@ -569,7 +598,7 @@ def two_site_count_stat(
 
 def sample_sets_to_bit_array(
     ts: tskit.TreeSequence, sample_sets: List[List[int]]
-) -> Tuple[np.ndarray, BitSet]:
+) -> Tuple[np.ndarray, np.ndarray, BitSet]:
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply python/tests/test_ld_matrix.py:601]: Not worth getting rid of the typehints now you've gone to the trouble of adding them. I'd just favour simplicity and readability over mypy-correctness

File: python/tests/test_ld_matrix.py (hunk 2 above)
Comment (jeromekelleher) [reply python/tests/test_ld_matrix.py:749]: Could just document one of the summary funcs, and leave the rest clear? It is helpful documenting the first one here

File: python/tests/test_ld_matrix.py [hunk 4]
```diff
@@ -699,6 +735,210 @@ def r2_summary_func(
...
+        p_Ab = state[1, k] / n
+        p_aB = state[2, k] / n
+
+        p_A = p_AB + p_Ab
+        p_B = p_AB + p_aB
+
+        D = p_AB - (p_A * p_B)
+        denom = p_A * p_B * (1 - p_A) * (1 - p_B)
+
+        if denom == 0 and D == 0:
```
Comment (jeromekelleher) [inline python/tests/test_ld_matrix.py:836]: See point above about NaN results. We found for the other stats API that it was easier to just let the division by zero happen and return NaN where the stat is undefined. Using a capturewarnings stops annoying warnings happening.

File: python/tskit/trees.py [hunk 5]
```diff
@@ -7567,6 +7567,65 @@ def __one_way_sample_set_stat(
...
+        sample_sets,
+        sites=None,
+        mode=None,
+    ):
+        if sample_sets is None:
+            sample_sets = self.samples()
+        if sites is not None and any(
+            not hasattr(a, "__getitem__") or isinstance(a, str) for a in sites
+        ):
+            raise ValueError("sites must be a list of lists, tuples, or ndarrays")
```
Comment (jeromekelleher) [inline python/tskit/trees.py:7582]: Need to test this

**Review by jeromekelleher — COMMENTED** [review]
LGTM - just needs a bit of extra testing on the input value checking.

We should open an issue to track dealing with undefined values,  if you disagree with letting the division by zero happen and returning NaN

File: c/tskit/trees.c (hunk 1 above)
Comment (apragsdale) [reply c/tskit/trees.c:3641]: That's a good question. Should D' (and other ratio statistics like r and r^2) return zero if there is no variation, or NaN? I had thought those results should be zero at first, but now I'm not so sure... We could tackle this in a follow-up issue, like you say.

**Review by apragsdale — COMMENTED** [review]
Thanks @lkirk! Great to see this moving along!

**Comment by lkirk** [comment]
re: docstrings, there's already one present in the code for `r2`, so I left that one in place but got rid of all of the ones in this PR.

Input validation testing is done, all testable branches are covered.

I'll open an issue for further NaN discussions.

**Comment by lkirk** [comment]
I'm a little confused by the results of the coverage suite. It's annotating lines on the diff, but the patch coverage is reported as 100%.

I've opened #2907 to further investigate the NaN behavior.

Squashed down preemptively, but let me know if there's anything else needed here.

File: python/tests/test_ld_matrix.py [hunk 6]
```diff
@@ -798,36 +993,100 @@ def get_all_site_partitions(n):
...
+
+@pytest.mark.parametrize(
+    "ts",
+    [
+        ts
+        for ts in get_example_tree_sequences()
+        if ts.id not in {"no_samples", "empty_ts"}
+    ],
+)
+@pytest.mark.parametrize("stat", SUMMARY_FUNCS.keys())
```
Comment (jeromekelleher) [inline python/tests/test_ld_matrix.py:1067]: Nice test! I didn't think of the trick of filtering by ID for the outliers :+1:

**Review by jeromekelleher — APPROVED** [review]
Thanks @lkirk, looks great. CodeCov is clearly lying, so let's merge rather than try to figure out what it's confused about.
